rfcs/2024-08-21-motion-presets.md

```diff
@@ -0,0 +1,555 @@
+---
+Start Date: 21-08-2024
+RFC PR: (leave this empty if no PR yet)
+Blade Issue: (leave this empty if no issue yet)
+---
+
+# Motion Presets RFC <!-- omit in toc -->
+
+### Table Of Contents <!-- omit in toc -->
+
+- [Summary](#summary)
+- [Basic Example](#basic-example)
+- [Motivation](#motivation)
+- [Detailed Design](#detailed-design)
+- [Drawbacks/Constraints](#drawbacksconstraints)
+- [Alternatives](#alternatives)
+- [Adoption strategy](#adoption-strategy)
+- [How do we educate people?](#how-do-we-educate-people)
+- [Open Questions](#open-questions)
+- [References](#references)
+
+# Summary
+
+# Basic Example
+
+```jsx
+import { Fade, Card, CardBody } from '@razorpay/components/blade';
+
+<Fade isVisible={showCard}>
+  <Card>
+    <CardBody>{/* Blade Card */}</CardBody>
+  </Card>
+</Fade>;
+```
+
+Checkout full [API Decisions for Motion Presets](#api-decisions)
+
+# Motivation
+
+You might have seen our previous [RFC on Motion Foundations](./2022-03-22-motion-rfc.md) we wrote in 2022. In that RFC we defined token structure and foundational principles for motion.
+
+As our org grows, we believe our products need to go a bit beyond the functionality. How do we make sure that our consumers really love our products? How do we add that cherry on top, those sprinkles, that pineapple on pizza?
+
+<img src="./images/motion-presets/motion-presets-meme.png" height="200px" /> <img src="./images/motion-presets/motion-presets-consumers-love.gif" height="200px" />
+
+To solve for consumer delight, the blade team is working on simplifying adding motion in our consumer applications and introduce consistent animations to our product.
+
+You can check our detailed design proposal on [Motion Refresh](https://docs.google.com/document/d/1Vj2E0v2IdLSHuJNRrmX7X-BFSqGpXMFU0jj7LLfoZGk/edit?usp=sharing) (Only accessible to Razorpay Employees)
+
+# Detailed Design
+
+For building presets, we have to figure out few things like
+
+- [API Decision for Motion Presets](#api-decisions)
+- [Libarary to use for Animations](#library-comparison)
+
+## API Decisions
+
+> [!NOTE]
+>
+> The API decisions here are only there to give some basic idea on the structure and usage. More accurate props will be updated here later once they are finalised in design.
+
+### Entry / Exit Animation Presets
+
+#### Fade
+
+<table>
+<tr>
+  <td>
+
+```jsx
+import { Fade } from '@razorpay/blade/components';
+
+<Fade isVisible={showCard}>
+  <Card>
+    <CardBody>
+      <Text>Fade In/Out Card</Text>
+    </CardBody>
+  </Card>
+</Fade>;
+```
+
+```ts
+type FadeProps = {
+  /**
+   * whether to animate entry and / or exit
+   *
+   * @default inout
+   */
+  variant: 'in' | 'out' | 'inout';
+
+  /**
+   * What should trigger the motion
+   *
+   * @default ['mount']
+   */
+  motionTriggers: ('mount' | 'hover' | 'tap' | 'inView')[];
```

Comment:
Why do we need to define tap and inview triggers? 

Consumers can do this on their end also like so: 

```
const isInView = useIsInView()

<Fade isVisible={isInView} />
```

Also if we are supporting tap, then we need to support click handlers too?

Reply:
I added all the triggers here which are natively supported by framer-motion so API is simpler

https://www.framer.com/motion/scroll-animations/#scroll-triggered-animations

> Also if we are supporting tap, then we need to support click handlers too?

Nope. Click handlers can be added to children directly no if they support it? This is purely for animation component. It shouldn't be used to add any logic.

E.g. I can just add click handlers on children if needed for some logic

```jsx
<Fade>
  <Card onClick={() => } />
</Fade>
```